File: Data Structures & Algorithms/anagram-groups/submission-2.ts
class Solution {
    /**
     * @param {string[]} strs
     * @return {string[][]}
     */
    groupAnagrams(strs: string[]): string[][] {
        // map sorted word to an array of words
        // check if map contains current word sorted
        // map get array and add new word or default to empty array and add new word
        const map : Map<string, string[]> = new Map<string, string[]>();
        

        for(let i = 0; i < strs.length; i++) {
            const sortedWord : string  = strs[i].split('').sort().join('');
            let wordArray : string[] = map.get(sortedWord) ?? [];
            wordArray.push(strs[i])
            map.set(sortedWord, wordArray)
        }

        // iterate over keyset and build out ans
        let ans: string[][] = []
        for(const str of map.keys()) {
            ans.push(map.get(str))
        }
        return ans;
    }
}

File: Data Structures & Algorithms/duplicate-integer/submission-2.py
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
        # to check if an array contains a duplicate, we instantiate an empty Number Set
        # we iterate through the array, if the current number is in number set then array contains duplicate, return true
        # otherwise, add number to set
        # return false

        numSet = set()

        for num in nums:
            if (num in numSet):
                return True
            numSet.add(num)
        
        return False